library/Fission/CLI/Pin.hs

```diff
@@ -45,65 +44,29 @@ run :: MonadRIO          cfg m
     => Has IPFS.BinPath  cfg
     => Has IPFS.Timeout  cfg
     => CID
-    -> UserConfig
+    -> NonEmpty IPFS.Peer
+    -> BasicAuthData
     -> m (Either SomeException CID)
--- run cid@(CID hash) userConfig = Exception.handleWith_ CLI.Error.put' $ do
-run cid@(CID hash) userConfig = do -- runExceptT do
-  -- Client.Runner runner <- Config.get
-
-  -- let
-  --   maybePeer = headMaybe $ peers userConfig
-
-  -- peer <- return $ maybeToEither maybePeer (toException "ERROR")
-
-  -- result  <- liftE $ IPFS.Peer.connect peer
-  -- hold <- liftE $ pin runner userConfig cid
-  -- CLI.Success.live hash
-  -- return $ Right cid
-
-  Client.Runner runner <- Config.get
-
+run cid@(CID hash) peers auth = runExceptT do
   logDebug $ "Remote pinning " <> display hash
+  Client.Runner runner <- Config.get
 
-  -- IPFS.Peer.connect IPFS.Peer.fission
-
-  let
-    mayPeer = headMaybe $ peers userConfig
-
-  peer <- ExceptT $ return $ maybeToEither mayPeer (toException MyError)
-  return $ IPFS.Peer.connect peer
-  bar <- liftE $ pin runner userConfig cid
-  -- CLI.Success.live hash
-  -- return cid
-
-  -- case foo' of
-  --   Left err -> return $ Left err
-  --   Right pr -> do
-  --     IPFS.Peer.connect pr
-
-  --     bar <- pin runner userConfig cid
+  liftE $ IPFS.Peer.connect IPFS.Peer.fission
```

Comment:
It's kind of hard to see in the diff, so here it is in plain:

```haskell
run :: MonadRIO          cfg m
    => MonadUnliftIO         m
    => HasLogFunc        cfg
    => HasProcessContext cfg
    => Has Client.Runner cfg
    => Has IPFS.BinPath  cfg
    => Has IPFS.Timeout  cfg
    => CID
    -> NonEmpty IPFS.Peer
    -> BasicAuthData
    -> m (Either SomeException CID)
run cid@(CID hash) peers auth = runExceptT do
  logDebug $ "Remote pinning " <> display hash
  Client.Runner runner <- Config.get

  liftE $ IPFS.Peer.connect IPFS.Peer.fission

  IPFS.Peer.connect $ head $ peers
  liftE $ pin runner cid auth

  CLI.Success.live hash
  return cid
```

Reply:
Breaking down a few of the changes:

# 1. `NonEmpty`

We shouldn't get this far into the function if we have no peers. In general the pattern in Haskell is to push any ambiguities as close to the application boundary as possible. IMO this is a good pattern to apply to any language. Here we do this here: https://github.com/fission-suite/web-api/pull/174/files#diff-07ef7e40bb53d7443ccd27f6e7dbe01dR69-R80 , and make it a non-empty list (`NonEmpty a`)

# 2. No More `Maybe`

Thanks to that `NonEmpty`, we no longer need to pattern match on `maybeHead`

# 3. `ExceptT`

Setup: Change the monad that we're running in to do automated error handling by sticking a `runExcepT` at the beginning. Now `>>=` uses the instance from `ExceptT` because it's the outermost one.

Payoff: Take anything in the form `m (Either err val)` and slap a `liftE` on the front. This does the railroad programming / error handling for us.